flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala

```diff
@@ -50,13 +53,29 @@ class TableSourceTable[T](
     this(tableSource, isStreamingMode, statistic, None, catalogTable)
   }
 
+  Preconditions.checkNotNull(tableSource)
+  Preconditions.checkNotNull(statistic)
+  Preconditions.checkNotNull(catalogTable)
+
+  val watermarkSpec: Option[WatermarkSpec] = catalogTable
+    .getSchema
+    .getWatermarkSpecs.asScala.headOption
+
+  if (TableSourceValidation.hasRowtimeAttribute(tableSource) && watermarkSpec.isDefined) {
+    throw new TableException(
+      "Watermark syntax in DDL can't work together with DefinedRowtimeAttributes interface. " +
+        "If watermark is specified in DDL, the underlying TableSource of connector shouldn't " +
+        "implement DefinedRowtimeAttributes.")
+  }
+
   // TODO implements this
   // TableSourceUtil.validateTableSource(tableSource)
 
   override def getRowType(typeFactory: RelDataTypeFactory): RelDataType = {
     TableSourceUtil.getRelDataType(
```

Comment:
This is wrong, `TableSourceTable`'s return type should not be decided by `TableSource`Util